driver-core/src/main/com/mongodb/operation/FindAndUpdateOperation.java

```diff
@@ -373,40 +372,34 @@ public Collation getCollation() {
-                validateCollationAndRetryWrites(connection, collation, retryWrites);
-                return executeWrappedCommandProtocol(binding, namespace.getDatabaseName(), getCommand(connection.getDescription()),
-                        getFieldNameValidator(), txnNumber, CommandResultDocumentCodec.create(decoder, "value"),
-                        connection, FindAndModifyHelper.<T>transformer());
-            }
-        });
+        return executeRetryableCommand(binding, retryWrites, namespace.getDatabaseName(), getFieldNameValidator(),
+                CommandResultDocumentCodec.create(decoder, "value"), new ConnectionCallable<BsonDocument>() {
+                    @Override
+                    public BsonDocument call(final Connection connection) {
+                        validateCollationAndRetryWrites(connection, collation, retryWrites);
+                        return getCommand(binding.getSessionContext(), connection.getDescription());
+                    }
+                }, FindAndModifyHelper.<T>transformer());
     }
 
     @Override
-    public void executeAsync(final AsyncWriteBinding binding, final SingleResultCallback<T> originalCallback) {
-        retryableWithConnection(binding, retryWrites, originalCallback, new AsyncCallableWithConnectionAndTransactionNumber<T>() {
-            @Override
-            public void call(final AsyncConnection connection, final Long txnNumber, final SingleResultCallback<T> callback) {
-                validateCollationAndRetryWrites(connection, collation, retryWrites, new AsyncCallableWithConnection() {
+    public void executeAsync(final AsyncWriteBinding binding, final SingleResultCallback<T> callback) {
+        executeRetryableCommand(binding, retryWrites, namespace.getDatabaseName(), getFieldNameValidator(),
+                CommandResultDocumentCodec.create(decoder, "value"),
+                new NestedCallable<AsyncConnection, NestedCallable<BsonDocument, Throwable>>() {
                     @Override
-                    public void call(final AsyncConnection connection, final Throwable t) {
-                        if (t != null) {
-                            callback.onResult(null, t);
-                        } else {
-                            executeWrappedCommandProtocolAsync(binding, namespace.getDatabaseName(),
-                                    getCommand(connection.getDescription()), getFieldNameValidator(), txnNumber,
-                                    CommandResultDocumentCodec.create(decoder, "value"), connection,
-                                    FindAndModifyHelper.<T>transformer(), callback);
-                        }
+                    public void call(final AsyncConnection connection, final NestedCallable<BsonDocument, Throwable> commandCallable) {
+                        validateCollationAndRetryWrites(connection, collation, retryWrites, new AsyncCallableWithConnection() {
```

Comment:
This struck me as more complicated than necessary by making something that could be synchronous -- validating the arguments -- and having it be asynchronous, which adds another chain.

I think instead we can make the sync and async code paths more similar by just requiring each to validate and create the command synchronously.  Something more like this:

    @Override
    public T execute(final WriteBinding binding) {
        return executeRetryableCommand(binding, retryWrites, namespace.getDatabaseName(), getFieldNameValidator(),
                CommandResultDocumentCodec.create(decoder, "value"),
                new CommandCreator() {
                    @Override
                    public BsonDocument create(final ConnectionDescription connectionDescription) {
                        validateCollationAndRetryWrites(connectionDescription, collation, retryWrites);
                        return getCommand(binding.getSessionContext(), connectionDescription);
                    }
                }, FindAndModifyHelper.<T>transformer());
    }

    @Override
    public void executeAsync(final AsyncWriteBinding binding, final SingleResultCallback<T> callback) {
        executeRetryableCommand(binding, retryWrites, namespace.getDatabaseName(), getFieldNameValidator(),
                CommandResultDocumentCodec.create(decoder, "value"),
                new CommandCreator() {
                    @Override
                    public BsonDocument create(final ConnectionDescription connectionDescription) {
                        validateCollationAndRetryWrites(connectionDescription, collation, retryWrites);
                        return getCommand(binding.getSessionContext(), connectionDescription);
                    }
                }, FindAndModifyHelper.<T>transformer(), callback);
    }

and requiring the async version of executeRetryableCommand to catch any exception thrown by the CommandCreator and call the callback with it.

Reply:
Done.